Invalid suggestion to consider borrowing `Option`'s content in tuple

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur. 
-->

Given the following code: ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7733e50640b6be7c6075f1aa5fab80f4)) <!-- Please provide a link to play.rust-lang.org -->

```rust
fn foo(a: &Option<String>, b: &Option<String>) {
    match (a, b) {
        (None, &c) | (&c, None) => &c.unwrap(),
        (&Some(ref c), _) => c,
    };
}
```

The current output is:

```
error[E0507]: cannot move out of a shared reference
 --> src/lib.rs:2:11
  |
2 |     match (a, b) {
  |           ^^^^^^ help: consider borrowing the `Option`'s content: `(a, b).as_ref()`
3 |         (None, &c) | (&c, None) => &c.unwrap(),
  |                 -
  |                 |
  |                 data moved here
  |                 move occurs because `c` has type `Option<String>`, which does not implement the `Copy` trait
```

The suggestion does not compile. Presumably it should suggest `(a.as_ref(), b.as_ref())`.

--- a/compiler/rustc_borrowck/src/diagnostics/move_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/move_errors.rs
@@ -218,18 +218,29 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
 
     fn report(&mut self, error: GroupedMoveError<'tcx>) {
         let (mut err, err_span) = {
-            let (span, use_spans, original_path, kind): (
+            let (span, use_spans, original_path, kind, has_complex_bindings): (
                 Span,
                 Option<UseSpans<'tcx>>,
                 Place<'tcx>,
                 &IllegalMoveOriginKind<'_>,
+                bool,
             ) = match error {
-                GroupedMoveError::MovesFromPlace { span, original_path, ref kind, .. }
-                | GroupedMoveError::MovesFromValue { span, original_path, ref kind, .. } => {
-                    (span, None, original_path, kind)
+                GroupedMoveError::MovesFromPlace {
+                    span,
+                    original_path,
+                    ref kind,
+                    ref binds_to,
+                    ..
                 }
+                | GroupedMoveError::MovesFromValue {
+                    span,
+                    original_path,
+                    ref kind,
+                    ref binds_to,
+                    ..
+                } => (span, None, original_path, kind, !binds_to.is_empty()),
                 GroupedMoveError::OtherIllegalMove { use_spans, original_path, ref kind } => {
-                    (use_spans.args_or_use(), Some(use_spans), original_path, kind)
+                    (use_spans.args_or_use(), Some(use_spans), original_path, kind, false)
                 }
             };
             debug!(
@@ -248,6 +259,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                             target_place,
                             span,
                             use_spans,
+                            has_complex_bindings,
                         ),
                     &IllegalMoveOriginKind::InteriorOfTypeWithDestructor { container_ty: ty } => {
                         self.cannot_move_out_of_interior_of_drop(span, ty)
@@ -290,6 +302,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         deref_target_place: Place<'tcx>,
         span: Span,
         use_spans: Option<UseSpans<'tcx>>,
+        has_complex_bindings: bool,
     ) -> DiagnosticBuilder<'a, ErrorGuaranteed> {
         // Inspect the type of the content behind the
         // borrow to provide feedback about why this
@@ -399,6 +412,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         let diag_name = self.infcx.tcx.get_diagnostic_name(def_id);
         if matches!(diag_name, Some(sym::Option | sym::Result))
             && use_spans.map_or(true, |v| !v.for_closure())
+            && !has_complex_bindings
         {
             err.span_suggestion_verbose(
                 span.shrink_to_hi(),

--- a/src/test/ui/suggestions/option-content-move-from-tuple-match.rs
+++ b/src/test/ui/suggestions/option-content-move-from-tuple-match.rs
@@ -1,0 +1,9 @@
+fn foo(a: &Option<String>, b: &Option<String>) {
+    match (a, b) {
+        //~^ ERROR cannot move out of a shared reference
+        (None, &c) => &c.unwrap(),
+        (&Some(ref c), _) => c,
+    };
+}
+
+fn main() {}

--- a/src/test/ui/suggestions/option-content-move-from-tuple-match.stderr
+++ b/src/test/ui/suggestions/option-content-move-from-tuple-match.stderr
@@ -1,0 +1,15 @@
+error[E0507]: cannot move out of a shared reference
+  --> $DIR/option-content-move-from-tuple-match.rs:2:11
+   |
+LL |     match (a, b) {
+   |           ^^^^^^
+LL |
+LL |         (None, &c) => &c.unwrap(),
+   |                 -
+   |                 |
+   |                 data moved here
+   |                 move occurs because `c` has type `Option<String>`, which does not implement the `Copy` trait
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0507`.

